fileuploader/large_file_upload_task.go

```diff
@@ -0,0 +1,149 @@
+type LargeFileUploadTask[T interface{}] interface {
+	Upload(progress ProgressCallBack) UploadResult[T]
+	Resume(progress ProgressCallBack) (UploadResult[T], error)
+	Cancel() error
+}
+
+// ByteStream is an interface that represents a stream of bytes
+type ByteStream interface {
+	io.Seeker
+	io.Reader
+	Stat() (os.FileInfo, error)
+}
+
+type largeFileUploadTask[T interface{}] struct {
+	uploadSession   UploadSession
+	adapter         abstractions.RequestAdapter
+	byteStream      ByteStream // *os.File by default implements ByteStream
+	maxSlice        int64
+	parsableFactory serialization.ParsableFactory
+	errorMappings   abstractions.ErrorMappings
+}
+
+func NewLargeFileUploadTask[T interface{}](adapter abstractions.RequestAdapter, uploadSession UploadSession, byteStream ByteStream, maxSlice int64, parsableFactory serialization.ParsableFactory, errorMappings abstractions.ErrorMappings) LargeFileUploadTask[T] {
+	return &largeFileUploadTask[T]{
+		adapter:         adapter,
+		uploadSession:   uploadSession,
+		byteStream:      byteStream,
+		maxSlice:        maxSlice,
+		parsableFactory: parsableFactory,
+		errorMappings:   errorMappings,
+	}
+}
+
+// UploadAsync uploads the byteStream in slices and returns the result of the upload
+func (l *largeFileUploadTask[T]) Upload(progress ProgressCallBack) UploadResult[T] {
+	result := NewUploadResult[T]()
+	var wg sync.WaitGroup
+	slices := l.createUploadSlices()
+	for _, slice := range slices {
+		wg.Add(1)
+		uploadSlice := slice
+		go func() {
+			defer wg.Done()
+			l.uploadAsync(progress, uploadSlice, result)
+		}()
+	}
+
+	wg.Wait()
+	return result
+}
+
+// Resume uploads the byteStream in slices and returns the result of the upload
+func (l *largeFileUploadTask[T]) Resume(progress ProgressCallBack) (UploadResult[T], error) {
+	// check if next expected ranges is empty
+
+	if len(l.uploadSession.GetNextExpectedRanges()) == 0 {
+		return nil, errors.New("UploadSession does not have next expected ranges")
+	}
+
+	if l.uploadSession.GetExpirationDateTime().After(time.Now()) {
+		return nil, errors.New("UploadSession has expired")
+	}
+
+	return l.Upload(progress), nil
+}
+
+// Cancel cancels the upload
+func (l *largeFileUploadTask[T]) Cancel() error {
+	requestInfo := abstractions.NewRequestInformationWithMethodAndUrlTemplateAndPathParameters(abstractions.DELETE, *l.uploadSession.GetUploadUrl(), make(map[string]string))
+	err := l.adapter.SendNoContent(context.Background(), requestInfo, l.errorMappings)
+	return err
+}
+
+func (l *largeFileUploadTask[T]) uploadAsync(progress ProgressCallBack, slice uploadSlice[T], result UploadResult[T]) {
+	maxRetry := 3
+	retry := 1
+	for retry < maxRetry {
+		// store the result of the upload
+		response, err := slice.UploadAsync(l.uploadSession, l.parsableFactory) // check if successful
+		if err != nil {
+			// if not successful, try again
+			if retry >= maxRetry {
+				result.SetUploadSucceeded(false)
+			}
+
+		} else {
+			result.SetUploadSucceeded(true)
```

Comment:
This looks to be set on each slice upload. Shouldn't it be set to true when the entire upload is completed?